modules/schema/src/main/java/org/apache/ignite/internal/schema/BinaryTupleParser.java

```diff
@@ -0,0 +1,215 @@
+     * Returns the binary tuple size in bytes.
+     */
+    public int size() {
+        return valueBase + getOffset(valueBase - entrySize);
+    }
+
+    /**
+     * Returns the number of elements in the tuple.
+     */
+    public int elementCount() {
+        return numElements;
+    }
+
+    /**
+     * Check if the binary tuple contains a null map.
+     */
+    public boolean hasNullMap() {
+        return entryBase > BinaryTupleSchema.HEADER_SIZE;
+    }
+
+    /**
+     * Get buffer positioned at the specified tuple element.
+     *
+     * If the element has NULL value then null is returned.
+     *
+     * @param index Index of the element.
+     * @return ByteBuffer with element bytes between current position and limit or null.
+     */
+    public ByteBuffer element(int index) {
+        var sink = new Sink() {
+            int begin, end;
+
+            @Override
+            public void nextElement(int index, int begin, int end) {
+                this.begin = begin;
+                this.end = end;
+            }
+        };
+
+        fetch(index, sink);
+        if (sink.begin == 0) {
+            return null;
+        }
+
+        return buffer.asReadOnlyBuffer().position(sink.begin).limit(sink.end).order(ByteOrder.LITTLE_ENDIAN);
+    }
+
+    /**
+     * Locate the specified tuple element.
+     *
+     * @param index Index of the element.
+     * @param sink Receiver.
+     */
+    public void fetch(int index, Sink sink) {
+        assert index >= 0;
+        assert index < numElements;
+
+        int entry = entryBase + index * entrySize;
+
+        int offset = valueBase;
+        if (index > 0) {
+            offset += getOffset(entry - entrySize);
+        }
+
+        int nextOffset = valueBase + getOffset(entry);
+        assert nextOffset >= offset;
+
+        if (offset == nextOffset && hasNullMap()) {
+            int nullIndex = BinaryTupleSchema.HEADER_SIZE + index / 8;
+            byte nullMask = (byte) (1 << (index % 8));
+            if ((buffer.get(nullIndex) & nullMask) != 0) {
+                sink.nextElement(index, 0, 0);
+                return;
+            }
+        }
+
+        sink.nextElement(index, offset, nextOffset);
+    }
+
+    /**
+     * Feeds the receiver with all tuple elements.
+     *
+     * @param sink Receiver.
+     */
+    public void parse(Sink sink) {
+        int entry = entryBase;
+        int offset = valueBase;
+
+        for (int i = 0; i < numElements; i++) {
+            int nextOffset = valueBase + getOffset(entry);
+            assert nextOffset >= offset;
+
+            if (offset == nextOffset && hasNullMap()) {
+                int nullIndex = BinaryTupleSchema.HEADER_SIZE + i / 8;
+                byte nullMask = (byte) (1 << (i % 8));
+                if ((buffer.get(nullIndex) & nullMask) != 0) {
+                    sink.nextElement(i, 0, 0);
+                    continue;
+                }
+            }
+
+            sink.nextElement(i, offset, nextOffset);
```

Comment:
In the fetch method you use it to return a pair of `begin` and `end` offsets.
Here, Sink semantic is different. You use it as a Consumer.

I think Sink could be a top-level public class with the consumer semantic,
and fetch() method can be rewritten to return an object with pair of values, e.g. Element is a good name.